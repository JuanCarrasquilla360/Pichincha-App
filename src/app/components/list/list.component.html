<div *ngIf="isLoading; else content">
    <div class="loader skeleton"></div>
</div>
<ng-template #content>
    <div class="search-container">
        <mat-form-field appearance="outline">
            <mat-label>Search</mat-label>
            <input matInput (keyup)="applyFilter($event)" placeholder="name" #input>
        </mat-form-field>
        <button class="test-agregar-btn" mat-raised-button (click)="navigateToAdd()">Agregar</button>
    </div>
    <div class="mat-elevation-z8 mat-table-responsive">
        
        <table mat-table [dataSource]="dataSource">
    
            <!-- Logo Column -->
            <ng-container matColumnDef="logo">
                <th mat-header-cell *matHeaderCellDef> Logo </th>
                <td mat-cell *matCellDef="let element"> <img class="logo" [src]="element.logo"></td>
            </ng-container>
    
            <!-- Name Column -->
            <ng-container matColumnDef="name">
                <th mat-header-cell *matHeaderCellDef> Nombre </th>
                <td mat-cell *matCellDef="let element"> {{element.name}} </td>
            </ng-container>
    
            <!-- Description Column -->
            <ng-container matColumnDef="description">
                <th mat-header-cell *matHeaderCellDef> <div class="icon-container"> Descripción <mat-icon class="icon" aria-hidden="false" aria-label="info" fontIcon="info"></mat-icon></div></th>
                <td mat-cell *matCellDef="let element"> {{element.description}} </td>
            </ng-container>
    
            <!-- Date Release Column -->
            <ng-container matColumnDef="date_release">
                <th mat-header-cell *matHeaderCellDef> <div class="icon-container">  Fecha de Liberación <mat-icon class="icon" aria-hidden="false" aria-label="info" fontIcon="info"></mat-icon></div></th>
                <td mat-cell *matCellDef="let element"> {{element.date_release}} </td>
            </ng-container>
    
            <!-- Date Revision Column -->
            <ng-container matColumnDef="date_revision">
                <th mat-header-cell *matHeaderCellDef> <div class="icon-container"> Fecha de Reestructuración <mat-icon class="icon" aria-hidden="false" aria-label="info" fontIcon="info"></mat-icon></div> </th>
                <td mat-cell *matCellDef="let element"> {{element.date_revision}} </td>
            </ng-container>
            <ng-container matColumnDef="actions">
                <th mat-header-cell *matHeaderCellDef> Acciones </th>
                <td mat-cell *matCellDef="let element">  
                  <button [matMenuTriggerFor]="menu" mat-icon-button>
                    <mat-icon>more_vert</mat-icon>
                  </button>
                  
                  <mat-menu #menu="matMenu">
                    <button mat-menu-item (click)="editProduct(element)">
                      <mat-icon>edit</mat-icon>
                      Editar
                    </button>
              
                    <button mat-menu-item (click)="deleteProduct(element)">
                      <mat-icon>delete</mat-icon>
                      Eliminar
                    </button>
                  </mat-menu>
              
                </td>
              </ng-container>
    
            <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
            <tr mat-row *matRowDef="let row; columns: displayedColumns;"></tr>
        </table>
    
        <mat-paginator [pageSizeOptions]="[5, 10, 20]" showFirstLastButtons aria-label="Select page of periodic elements">
        </mat-paginator>
    </div>
</ng-template>