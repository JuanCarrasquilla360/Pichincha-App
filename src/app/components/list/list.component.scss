.logo {
    width: 80px;
    margin-top: 3px;
}

.icon {
    color: #BBB;
    position: absolute;
    bottom: -4px;
    margin-left: 5px;
}
.mat-table-responsive {
    overflow: scroll;
}

.icon-container {
    position: relative;
}

.loader {
    width: 100%;
    height: 500px;
}

.skeleton {
    background: #aaa;
    border-radius: 4px;
    animation: pulse 1.2s infinite ease-in-out;
}

@keyframes pulse {
    0% {
        opacity: 1;
    }

    50% {
        opacity: 0.4;
    }

    100% {
        opacity: 1;
    }
}

.mdc-button{
    background-color: #ffdd03 !important;
    color: #142559 !important;
    font-weight: bold;
    height: 56px !important;
}

.search-container {
    display: flex;
    justify-content: space-between;
}

.hide {
    display: none;
}