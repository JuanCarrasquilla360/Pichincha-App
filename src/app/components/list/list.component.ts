import { AfterViewInit, Component, Inject, PLATFORM_ID, ViewChild } from '@angular/core';
import { CommonModule, isPlatformBrowser } from '@angular/common';
import { MatPaginator, MatPaginatorModule } from '@angular/material/paginator';
import { MatTableDataSource, MatTableModule } from '@angular/material/table';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { MatIconModule } from '@angular/material/icon';
import { MatGridListModule } from '@angular/material/grid-list';
import moment from 'moment';
import { deleteProduct, getProducts } from '../../services/product.service';
import { ProductsModel } from '../../models/products.model';
import { MatButtonModule } from '@angular/material/button';
import { MatMenuModule } from '@angular/material/menu';
import { MatDialogModule, MatDialog } from '@angular/material/dialog';
import { MatSnackBarModule, MatSnackBar } from '@angular/material/snack-bar';
import { Router } from '@angular/router';
import { DialogComponent } from '../dialog/dialog.component';


@Component({
  selector: 'app-list',
  standalone: true,
  imports: [
    CommonModule,
    MatTableModule,
    MatPaginatorModule,
    MatFormFieldModule,
    MatInputModule,
    MatIconModule,
    MatGridListModule,
    MatButtonModule,
    MatMenuModule,
    MatSnackBarModule,
    MatDialogModule
  ],
  templateUrl: './list.component.html',
  styleUrl: './list.component.scss'
})
export class ListComponent implements AfterViewInit {
  displayedColumns: string[] = ['logo', 'name', 'description', 'date_release', 'date_revision', "actions"];
  dataSource = new MatTableDataSource<ProductsModel>([]);
  rowsData: ProductsModel[] = []
  isLoading: boolean = false

  @ViewChild(MatPaginator) paginator!: MatPaginator;

  constructor(private router: Router, @Inject(PLATFORM_ID) private platformId: Object, public dialog: MatDialog, private snackBar: MatSnackBar) {
    this.isLoading = true
    getProducts().then(data => {
      this.rowsData = this.formattedData(data)
      this.dataSource = new MatTableDataSource<ProductsModel>(this.rowsData);
      this.isLoading = false
      if (isPlatformBrowser(this.platformId)) {
        localStorage.setItem('products', JSON.stringify(data));
      }
    })
  }

  navigateToAdd() {
    this.router.navigate(['/agregar-producto']);
  }


  ngAfterViewInit() {
    setTimeout(() => {
      this.dataSource.paginator = this.paginator;
    }, 250);
  }

  applyFilter(event: Event) {
    const filterValue = (event.target as HTMLInputElement).value;
    this.dataSource.filter = filterValue.trim().toLowerCase();
  }

  formattedData(arr: ProductsModel[]) {
    return arr.map(data => ({
      ...data,
      date_release: moment(data.date_release).format('MM/DD/YYYY'),
      date_revision: moment(data.date_revision).format('MM/DD/YYYY')
    }))
  }
  deleteProduct(e: any) {
    this.openConfirmDialog(e.name).afterClosed().subscribe(res => {     
      if (res) {
        deleteProduct(e.id).then(()=>{
          this.dataSource.data = this.dataSource.data.filter(p => p !== e)
          this.dataSource._updateChangeSubscription();  
          this.snackBar.open("Producto eliminado exitosamente", undefined, {
            duration: 2000, verticalPosition: "bottom"
          })
        })
      }
    });

  }
  editProduct(e: any) {
    this.router.navigate([`/agregar-producto/${e.id}`]);
  }

  openConfirmDialog(productName: string) {
    return this.dialog.open(DialogComponent, { width: "450px", data: {productName} })
  }


}
