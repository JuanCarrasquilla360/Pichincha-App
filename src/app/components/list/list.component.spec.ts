import { ComponentFixture, TestBed, } from '@angular/core/testing';
import { ListComponent } from './list.component';
import { Router } from '@angular/router';

describe('ListComponent', () => {
  let component: ListComponent;
  let fixture: ComponentFixture<ListComponent>;
  let compiled: HTMLElement;
  let router: Router;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ListComponent],
    })
      .compileComponents();
    fixture = TestBed.createComponent(ListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    router = TestBed.inject(Router);
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should match with snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

});
