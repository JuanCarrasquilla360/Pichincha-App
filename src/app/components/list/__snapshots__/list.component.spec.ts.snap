// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListComponent should match with snapshot 1`] = `
<div
  id="root1"
>
  <div>
    <div
      class="loader skeleton"
    />
  </div>
  
  
</div>
`;
